src/ucp/core/ucp_worker.c

```diff
@@ -2514,19 +2554,25 @@ void ucp_worker_keepalive_add_ep(ucp_ep_h ep)
 /* EP is removed from worker */
 void ucp_worker_keepalive_remove_ep(ucp_ep_h ep)
 {
-    ucp_worker_h worker = ep->worker;
-    ucp_ep_ext_gen_t *ep_ext;
+    ucp_worker_h worker   = ep->worker;
+    ucs_list_link_t *iter = worker->keepalive.iter;
 
     if (!ucp_worker_keepalive_is_enabled(worker)) {
         ucs_assert(worker->keepalive.iter == &worker->all_eps);
         return;
     }
 
-    /* EP is removed from worker all_eps, so we can't use list->next value */
+    ucs_assert(!ucs_list_is_empty(&worker->all_eps));
+
     if (worker->keepalive.iter == &ucp_ep_ext_gen(ep)->ep_list) {
-        ep_ext = ucs_list_next(worker->keepalive.iter,
-                               ucp_ep_ext_gen_t, ep_list);
-        worker->keepalive.iter = &ep_ext->ep_list;
+        /* if iterator points into EP to be removed - then
```

Comment:
1. add new function ucs_list_is_only(head, elem)
2. in here:
if (ucs_list_is_only) {
  worker->iter = &all_eps;
} else of (worker->iter == &ep->list) {
  next_ep(worker)
}

Reply:
added ucs_list_is_only, updated remove_ep function